gpus/python-gpu-example.html.md

```diff
@@ -0,0 +1,242 @@
+# initial_size = "50gb"            # Uncomment to set the size for the volume created on first deployment
+```
+
+### The Docker image
+
+Ready-made Docker images exist for many ML-related projects. There is a CUDA-enabled Jupyter Docker image, but it's kind of a black box and it's not maintained by the Jupyter folks, though they link to it. 
+
+Here we'll go step by step from a small Ubuntu image, installing and configuring a Python development environment that can use the GPU's CUDA capabilities, kind of like we would with a normal computer.
+
+Here's a summary of what the Dockerfile does;
+
+* Uses Ubuntu 22.04 as a base image, as it's compatible with the NVIDIA drivers on Fly Machines.
+* Installs system-wide packages with the apt package manager: python3, python3-pip, python3-venv, python3-wheel, git, nano (substitute your favourite terminal-compatible text editor here).
+* Adds a non-root user to own the Python venv and run things. Gets the name for this user from a `[build.arg]` set in `fly.toml`.
+* Copies the scripts for root and the user to run at startup.
+* CMD invokes the first script with the non-root user name as argument.
+
+```Dockerfile
+FROM ubuntu:22.04
+RUN apt update -q && apt install -y python3 python3-pip python3-venv python3-wheel git nano && \
+    apt clean && rm -f /var/lib/apt/lists/*_*
+
+ARG NONROOT_USER
+RUN echo "User will be $NONROOT_USER"
+ENV PYTHON_USER=$NONROOT_USER
+
+# Create unprivileged user with a home dir and using bash
+RUN useradd -ms /bin/bash $PYTHON_USER
+
+COPY --chmod=0755 ./entrypoint.sh ./entrypoint.sh
+COPY --chown=$PYTHON_USER:$PYTHON_USER --chmod=0755 ./post-initialization.sh ./post-initialization.sh
+# If you have a requirements.txt for the project, uncomment this and
+# adjust post-initialization.sh to use it
+# COPY --chown=$PYTHON_USER:$PYTHON_USER requirements.txt .
+
+# CMD ["sleep", "inf"]
+CMD ["/bin/bash", "-c", "./entrypoint.sh $PYTHON_USER"]
+```
+
+### Entrypoint script
+When this Machine boots, it runs the script `entrypoint.sh` as root. 
+
+This script gives the non-root user (whose username is set to `pythonuser` via a build argument in `fly.toml`) ownership of the non-root user's home directory to that user. This is necessary, because we're mounting a Fly Volume over that point in the Machine file system.
+
+Then it runs `nvidia-smi` to make sure the GPU drivers are loaded and the device is ready for the non-root user to use.
+
+Then it runs the next script (`post-initialization.sh`) as the non-root (`pythonuser`) user.
+
+You could instead have it run `sleep inf` here, and `fly ssh console` into the Machine after deployment to finish setting up the environment on the persistent volume.
+
+```bash
+#!/bin/bash
+
+USERNAME=$1
+
+echo "Inside entrypoint script."
+chown $USERNAME:$USERNAME /home/$USERNAME
+nvidia-smi # This seems to initialize the driver so that non-root user can use the GPU
```

Comment:
seems? :)

Reply:
That was an accurate description at the time lol